### PR TITLE
Action Subscriber 3.0 Proposal

### PR DESCRIPTION
**Pull request by mmmries**
Addresses #70 

The result of our recent explorations into thread safety issues has resulted in us finding out that the only safe place for us to send back message acknowledgements is within the consumer block that we register with `march_hare` (note: Bunny has slightly better concurrency guarantees, but we need to support both so we are using the lowest-common-denominator).

It turns out that the official rabbitmq drivers for C and Java both make this same recommendation of handling all messages form a single channel in a single thread. This means that for all `--mode=subscrbe` subscribers, shared threadpools managed by `action_subscriber` are off the table.

This PR is a proposal for a `3.0.0.pre1` version which will use `lifeguard` threadpools when in `pop` mode (polling), and will register multiple consumers (and support multiple connections) when in `subscribe` mode. What this means in essence is that instead drawing your routes like this:

```ruby
background_pool = ::ActionSubscriber::Threadpool.new_pool(:background, 10)
::ActionSubscriber.draw_routes do
  default_routes_for ::JobSubscriber
  route ::JobSubscriber, :created
    :acknowledgements => true,
    :prefetch => 10,
    :queue => "me.other_service.jobs.created_low",
    :threadpool => background_pool
end
```

You will now specify this as:

```ruby
::ActionSubscriber.draw_routes do
  default_routes_for ::JobSubscriber # this will now be a single threaded subscriber by default
  connection(:background, :threadpool_size => 10) do
    route ::JobSubscriber, :created,
      :acknowledgements => true,
      :concurrency => 10,
      :queue => "me.other_service.jobs.created_low",
  end
end
```

This allows you to control the concurrency level of each route, but also funnel the work into a shared threadpool (at the connection level).

> Note: for subscribers using the `pop` mode threadpools will still work just fine

Other changes proposed for 3.0:

* Remove all publisher code in favor of `active_publisher`
* Access to the connections is synchronized with a mutex (ie. ActionSubscriber::RabbitConnection.with_connect(:default){ |connection| do_some_work }`)
* Change `prefetch` to default value of 2 since each consumer will only do 1 job at at time now
* Remove the inferred routes fallback, clients must explicitly call `draw_routes` now
* Remove the attempted auto-loading of subscribers, we don't need this at all since we got rid of inferred routes

/cc @abrandoned @film42 @quixoten @brettallred @brianbroderick

### CODE REVIEW

**Comment by mmmries** [comment]
BTW in order to confirm that new concurrency model I ran a test subscriber under both jRuby and MRI that had a connection (with 20 threads) and registered a route with `:concurrency => 20`. The route sleeps for 10 seconds at a time so I could easily check the logs to see that it started 20 jobs at a time, then after 10 seconds it finished all 20 jobs.  Rabbit shows 20 consumers registered in both cases and 20 unacknowledged messages until the whole queue was worked through.

**Comment by brianbroderick** [comment]
@mmmries This all sounds great and we, at Nuvi, appreciate your efforts - we probably owe you lunch or something :).

Do you have any preliminary indications of how these proposed changes affect overall speed & efficiency?

**Comment by mmmries** [comment]
@brianbroderick the tests I've done so far don't show any change in performance between the two concurrency models. The application that I was testing spends between 5 and 50ms on each job, and it's throughput after these changes was the same as it was before these changes: ~150 jobs/sec on a single queue. In this case my application is the bottleneck so I'm not sure how much the performance of the library itself has changed.

I'll see if I can do some smaller benchmarks where a simplistic application can handle a lot more throughput to see how the performance has changed.

**Comment by abrandoned** [comment]
think this is a great step in the long-term success direction :shipit:

**Comment by film42** [comment]
Great PR! :shipit:

**Comment by mmmries** [comment]
I cut the pre-release gem from this branch because if I merge these changes into master, no one will be able to branch off of master for a while. `3.0.0.pre1` is released. Going to try to get it up into production in one of our apps today

**Comment by mmmries** [comment]
looks like I have some cases where the process hangs on exist, I'll make sure we are cleaning up the connections so we don't get broken builds

File: lib/action_subscriber/rabbit_connection.rb [hunk 1]
```diff
@@ -3,57 +3,40 @@
...
+      subscriber_connections.all?{|_name, connection| connection.connected?}
     end
 
-    def self.subscriber_connection
+    def self.subscriber_disconnect!
       SUBSCRIBER_CONNECTION_MUTEX.synchronize do
-        return @subscriber_connection if @subscriber_connection
-        @subscriber_connection = create_connection
+        subscriber_connections.each{|_name, connection| connection.close}
+        @subscriber_connections = []
```
Comment (abrandoned) [inline lib/action_subscriber/rabbit_connection.rb]: seems odd that we would set this to `{}` on initialization and then `[]1 on disconnect; maybe we should include concurrent-ruby and use Concurrent::Map in both cases?

File: lib/action_subscriber/rabbit_connection.rb [hunk 2]
```diff
@@ -3,57 +3,40 @@
...
-
-        @publisher_connection = nil
+        fail ArgumentError, "a #{name} connection already exists" if subscriber_connections[name]
+        subscriber_connections[name] = create_connection(settings)
       end
     end
 
     def self.subscriber_connected?
-      subscriber_connection.try(:connected?)
+      subscriber_connections.all?{|_name, connection| connection.connected?}
```
Comment (abrandoned) [inline lib/action_subscriber/rabbit_connection.rb]: should we be using the mutex here as well?

File: lib/action_subscriber/middleware/env.rb [hunk 3]
```diff
@@ -41,12 +41,14 @@ def initialize(subscriber, encoded_payload, properties)
...
 
       def acknowledge
+        fail ::RuntimeError, "you can't acknowledge messages under the polling API" unless @channel
         acknowledge_multiple_messages = false
         @channel.ack(@delivery_tag, acknowledge_multiple_messages)
         true
       end
 
       def reject
+        fail ::RuntimeError, "you can't acknowledge messages under the polling API" unless @channel
```
Comment (abrandoned) [inline lib/action_subscriber/middleware/env.rb]: probably language should be "reject" here?

File: lib/action_subscriber/rabbit_connection.rb (hunk 2 above)
Comment (mmmries) [inline lib/action_subscriber/rabbit_connection.rb]: switched to use the mutex here

File: lib/action_subscriber/rabbit_connection.rb (hunk 1 above)
Comment (mmmries) [inline lib/action_subscriber/rabbit_connection.rb]: fixed

**Comment by mmmries** [comment]
Just released `3.0.0.pre2`. We think it fixes a CPU utilization error under the `pop` mode with threadpools. Going to release it to a few of our internal services to test it under load.

**Comment by mmmries** [comment]
Here are the list of remaining pain points that I'm aware of:

* Test suite hangs at on jRuby sometimes (all tests completed, but the process doesn't exit cleanly)
* The API between the binstub => `ActionSubscriber::Babou` => `ActionSubscriber` is pretty messy. I really want to clean this up and have a solid setup and tear-down procedure.

**Comment by mmmries** [comment]
After 2 days of baking in production under load this is looking very stable. I'm going to go ahead and release the `3.0.0` version this morning. We can address the remaining painpoints in a `3.0.1`. I don't want to hold up people using the stable version of this library.

**Comment by liveh2o** [comment]
:shipit:

**Pull request merged**